src/value.h

```diff
@@ -85,159 +93,185 @@ class RangeType {
-	str_utf8_wrapper( size_t n, char c ) : std::string(n, c), cached_len(-1) { }
-	~str_utf8_wrapper() {}
-	
 	glong get_utf8_strlen() const {
-		if (cached_len < 0) {
-			cached_len = g_utf8_strlen(this->c_str(), this->size());
+		if (str_ptr->second < 0) {
+			str_ptr->second = g_utf8_strlen(this->str_ptr->first.c_str(), this->str_ptr->first.size());
 		}
-		return cached_len;
+		return str_ptr->second;
 	};
+
+
 private:
-	mutable glong cached_len;
+	str_ptr_t str_ptr;
 };
 
-
 class Value
 {
 public:
-	typedef std::vector<ValuePtr> VectorType;
-
-  enum class ValueType {
-    UNDEFINED,
-    BOOL,
-    NUMBER,
-    STRING,
-    VECTOR,
-    RANGE
-  };
-  static const Value undefined;
-
-  Value();
-  Value(bool v);
-  Value(int v);
-  Value(double v);
-  Value(const std::string &v);
-  Value(const char *v);
-  Value(const char v);
-  Value(const VectorType &v);
-  Value(const RangeType &v);
-  ~Value() {}
-
-  ValueType type() const;
-  bool isDefined() const;
-  bool isDefinedAs(const ValueType type) const;
-  bool isUndefined() const;
-
-  double toDouble() const;
-  bool getDouble(double &v) const;
-  bool getFiniteDouble(double &v) const;
-  bool toBool() const;
-  std::string toString() const;
-  std::string toString(const tostring_visitor *visitor) const;
-  std::string toEchoString() const;
-  std::string toEchoString(const tostring_visitor *visitor) const;
-  void toStream(std::ostringstream &stream) const;
-  void toStream(const tostream_visitor *visitor) const;
-  std::string chrString() const;
-  const VectorType &toVector() const;
-  bool getVec2(double &x, double &y, bool ignoreInfinite = false) const;
-  bool getVec3(double &x, double &y, double &z) const;
-  bool getVec3(double &x, double &y, double &z, double defaultval) const;
-  RangeType toRange() const;
+	typedef std::vector<Value> VectorType;
+	enum class ValueType {
+		UNDEFINED,
+		BOOL,
+		NUMBER,
+		STRING,
+		VECTOR,
+		RANGE
+	};
+	static const Value undefined;
+
+	class VectorPtr {
+	public:
+		VectorPtr() : ptr(make_shared<VectorType>()) {}
+		VectorPtr(double x, double y, double z);
+		VectorPtr(const VectorPtr &) = delete; // never copy, move instead
+		VectorPtr& operator=(const VectorPtr &) = delete; // never copy, move instead
+		VectorPtr(VectorPtr&&) = default; // default move constructor
+		VectorPtr& operator=(VectorPtr&&) = default; // default move assignment
+
+		// Copy explicitly only when necessary
+		// We can't use unique_ptr because of this :(
+		VectorPtr clone() const { VectorPtr c; c.ptr = this->ptr; return c; }
+
+		const Value::VectorType &operator*() const noexcept { return *ptr; }
+		Value::VectorType *operator->() const noexcept { return ptr.get(); }
+		// const accesses to VectorType require .clone to be move-able
+		const Value &operator[](size_t idx) const noexcept { return idx < ptr->size() ? (*ptr)[idx] : Value::undefined; }
+		// non-const return operator[] is only accessible from protected toVectorPtrRef
```

Comment:
... or from ```instance.clone()```

Reply:
Some sizeable changes were made to the interface of vectors in #3102.  VectorType iterators and `operator[]` *only* provide const reference access to elements now. 
This means even a non-const cloned vector can't move elements out of the original, which fits better with supposed immutability.

I guess there's still one remaining danger that a clone could have additional elements inserted into it, which would alter the original.
I'm not sure how to solve that without making vectors truly immutable upon construction, but this would require very complicated range-based constructors which can somehow forward arguments to `std::vector`, through `boost::variant`, etc.  I event spent a day thinking if this could be done, but I was unable to figure out.  Also related to @t-paul's [comment here](https://github.com/openscad/openscad/pull/2881#discussion_r336788449) about such difficulties.